mrjob/fs/gcs.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import fnmatch
+import logging
+
+# from mrjob.aws import s3_endpoint_for_region
+from mrjob.fs.base import Filesystem
+from mrjob.parse import urlparse
+from mrjob.runner import GLOB_RE
+
+
+from apiclient import discovery
+from oauth2client.client import GoogleCredentials
+from apiclient.http import MediaIoBaseDownload
+from apiclient import http
+import io
+import json
+import base64
+import binascii
+
+log = logging.getLogger(__name__)
+
+
+class GCSFilesystem(Filesystem):
+    """Filesystem for Google Cloud Storage (GCS) URIs. Typically you will get one of these via
+    ``DataprocJobRunner().fs``, composed with
+    :py:class:`~mrjob.fs.ssh.SSHFilesystem` and
+    :py:class:`~mrjob.fs.local.LocalFilesystem`.
+    """
+
+    def __init__(self):
+        credentials = GoogleCredentials.get_application_default()
+        self.service = discovery.build('storage', 'v1', credentials=credentials)
+
+    def can_handle_path(self, path):
+        return is_gcs_uri(path)
+
+    def du(self, path_glob):
+        """Get the size of all files matching path_glob."""
+        return sum(item[1] for item in self.lsWithInfo(path_glob))
+
+    def ls(self, path_glob):
+        for item in self.lsWithInfo(path_glob):
+            yield item[0]
+
+    def lsWithInfo(self, path_glob):
+        """Recursively list files on GCS and includes some metadata about them:
+        - uri
+        - size
+        - bucket name
+        - object name
+        - md5 hash
+        - etag
+
+        *path_glob* can include ``?`` to match single characters or
+        ``*`` to match 0 or more characters. Both ``?`` and ``*`` can match
+        ``/``.
+        """
+
+        scheme = urlparse(path_glob).scheme
+
+        # support globs
+        glob_match = GLOB_RE.match(path_glob)
+
+        # we're going to search for all keys starting with base_uri
+        if glob_match:
+            # cut it off at first wildcard
+            base_uri = glob_match.group(1)
+        else:
+            base_uri = path_glob
+
+        bucket_name, base_name = parse_gcs_uri(base_uri)
+
+        # allow subdirectories of the path/glob
+        if path_glob and not path_glob.endswith('/'):
+            dir_glob = path_glob + '/*'
+        else:
+            dir_glob = path_glob + '*'
+
+
+        fields_to_return = "nextPageToken,items(id,size,bucket,name,md5Hash,etag)"
```

Comment:
Is there any additional expense to requesting this info, if we don't use it? I'm guessing the API call was designed for us to only ask for what we need.